### PR TITLE
feat(swarm): expose `ConnectionId` and add conn duration metric

### LINKED ISSUE #3135: feat(swarm): expose duration of a connection
## Description

Expose the duration a connection was alive.

I would assume the easiest way to expose this is via `SwarmEvent::ConnectionClosed`.

https://github.com/libp2p/rust-libp2p/blob/7803524a76fb52817b00bcb61544be3a1a317ebd/swarm/src/lib.rs#L195-L207

## Motivation

Needed for monitoring purposes, e.g. to be exposed to Prometheus via libp2p-metrics.

## Are you planning to do it yourself in a pull request?

No

### PR DESCRIPTION
**Pull request by mxinden**
## Description

- Exposes the `ConnectionId` in the various `SwarmEvent` variants.
- Tracks connection duration in `libp2p-metrics::swarm`.

## Notes & open questions

~~Draft as it depends on https://github.com/prometheus/client_rust/pull/137.~~

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: misc/metrics/CHANGELOG.md [hunk 1]
```diff
@@ -3,7 +3,12 @@
 - Raise MSRV to 1.65.
   See [PR 3715].
 
+- Replace `libp2p_swarm_connections_closed` `Counter` with `libp2p_swarm_connections_duration` `Histogram` which additionally tracks the duration of a connection.
+  Note that you can use the `_count` metric of the `Histogram` as a replacement for the `Counter`.
+  See [PR XXX].
+
 [PR 3715]: https://github.com/libp2p/rust-libp2p/pull/3715
+[PR XXX]: https://github.com/libp2p/rust-libp2p/pull/XXX
```
Comment (thomaseizinger) [inline misc/metrics/CHANGELOG.md]:
```suggestion
  See [PR 3927].

[PR 3715]: https://github.com/libp2p/rust-libp2p/pull/3715
[PR 3927]: https://github.com/libp2p/rust-libp2p/pull/3927
```

File: swarm/CHANGELOG.md [hunk 2]
```diff
@@ -50,6 +50,10 @@
 - Remove deprecated `NetworkBehaviourAction` type.
   See [PR 3919].
 
+- Expose `ConnectionId` on `SwarmEvent::{ConnectionEstablished,ConnectionClosed,IncomingConnection,IncomingConnectionError,OutgoingConnectionError,Dialing}`.
```
Comment (thomaseizinger) [inline swarm/CHANGELOG.md:53]: How close are we to using `FromSwarm` in `SwarmEvent` (or the other way round)?

File: swarm/CHANGELOG.md [hunk 3]
```diff
@@ -62,6 +66,7 @@
 [PR 3886]: https://github.com/libp2p/rust-libp2p/pull/3886
 [PR 3912]: https://github.com/libp2p/rust-libp2p/pull/3912
 [PR 3919]: https://github.com/libp2p/rust-libp2p/pull/3919
+[PR XXX]: https://github.com/libp2p/rust-libp2p/pull/XXX
```
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]:
```suggestion
[PR 3927]: https://github.com/libp2p/rust-libp2p/pull/3927
```

File: swarm/CHANGELOG.md [hunk 4]
```diff
@@ -50,6 +50,10 @@
 - Remove deprecated `NetworkBehaviourAction` type.
   See [PR 3919].
 
+- Expose `ConnectionId` on `SwarmEvent::{ConnectionEstablished,ConnectionClosed,IncomingConnection,IncomingConnectionError,OutgoingConnectionError,Dialing}`.
+  Also emit `SwarmEvent::Dialing` for dials with unknown `PeerId`.
+  See [PR XXX].
```
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]:
```suggestion
  See [PR 3927].
```

File: swarm/src/lib.rs [hunk 5]
```diff
@@ -1111,10 +1135,12 @@ where
...
                 let peer_id = opts.get_or_parse_peer_id();
+                let connection_id = opts.connection_id();
                 if let Ok(()) = self.dial(opts) {
-                    if let Ok(Some(peer_id)) = peer_id {
-                        return Some(SwarmEvent::Dialing(peer_id));
-                    }
+                    return Some(SwarmEvent::Dialing {
+                        peer_id: peer_id.ok().flatten(),
+                        connection_id,
+                    });
```
Comment (thomaseizinger) [inline swarm/src/lib.rs:1143]: :+1:

**Review by thomaseizinger — COMMENTED** [review]
Direction looks good.

File: swarm/CHANGELOG.md (hunk 2 above)
Comment (mxinden) [reply swarm/CHANGELOG.md:53]: As far as I can tell, only the `Handler` in `ConnectionClosed` is a problem:

``` rust
/// [`FromSwarm`] variant that informs the behaviour about a closed connection to a peer.
///
/// This event is always paired with an earlier
/// [`FromSwarm::ConnectionEstablished`] with the same peer ID, connection ID
/// and endpoint.
#[allow(deprecated)]
pub struct ConnectionClosed<'a, Handler> {
    pub peer_id: PeerId,
    pub connection_id: ConnectionId,
    pub endpoint: &'a ConnectedPoint,
    pub handler: Handler,
    pub remaining_established: usize,
}
```

Comment (thomaseizinger) [reply swarm/CHANGELOG.md:53]: Right, that is more or less tracked in https://github.com/libp2p/rust-libp2p/issues/3046.

File: misc/metrics/src/swarm.rs [hunk 6]
```diff
@@ -261,25 +296,42 @@ impl<TBvEv, THandleErr> super::Recorder<libp2p_swarm::SwarmEvent<TBvEv, THandleE
...
     protocols: String,
 }
 
-type ConnectionEstablishmentDurationLabels = ConnectionEstablishedLabels;
-
 #[derive(EncodeLabelSet, Hash, Clone, Eq, PartialEq, Debug)]
 struct ConnectionClosedLabels {
-    role: Role,
-    protocols: String,
+    // TODO: Should be Option<ConnectionError>. Needs https://github.com/prometheus/client_rust/pull/137
```
Comment (thomaseizinger) [inline misc/metrics/src/swarm.rs]: Do we wait for this before merging here?

**Review by thomaseizinger — APPROVED** [review]

File: misc/metrics/src/swarm.rs (hunk 6 above)
Comment (mxinden) [reply misc/metrics/src/swarm.rs]: Oversight. Fixed with [9578315](https://github.com/libp2p/rust-libp2p/pull/3927/commits/9578315e6409a50693ab1ec746fd2c6b589f5166). Thanks for the catch Thomas!

**Pull request merged**